### PR TITLE
Parse headers as markdown (issues #69 #73)

### PR DESCRIPTION
**Pull request by jmanuel1**
Parse markdown headers to grab the wanted text, even if markdown is embedded within the header. Especially if the header has links or inline images.

See #73 for explanation.

### CODE REVIEW

File: lib/transform.js [hunk 1]
```diff
@@ -3,7 +3,8 @@
 var _             = require('underscore')
   , anchor        = require('anchor-markdown-header')
   , updateSection = require('update-section')
-  , getHtmlHeaders = require('./get-html-headers');
+  , getHtmlHeaders = require('./get-html-headers')
+  , md            = require('markdown-to-ast');
```
Comment (jez) [inline lib/transform.js]: Just curious (I haven't looked into this too much), what were some of the factors that led to this markdown parser? Seems like there are a number of other options: `marked-ast`, `markdown-ast`, `markdown-to-ast` to name a few. Was this one easier to use, faster, more well-documented...)

File: lib/transform.js [hunk 2]
```diff
@@ -24,64 +25,41 @@ function addAnchor(mode, header) {
...
+function getMarkdownHeaders (lines, maxHeaderLevel) {
+  function extractText (header) {
+    return header.children
+      .filter(function (x) {
+        return x.type !== 'Image';
+      })
+      .map(function (x) {
+        if (x.type === 'Link') {
+          return extractText(x);
+        }
```
Comment (jez) [inline lib/transform.js]: > ```javascript
> .map(function (x) {
>   if (x.type === 'Link') {
>     return extractText(x);
>   }
>   return x.raw;
> })
> ```

Wouldn't we want to recurse all the way down to the leaves? i.e.

```javascript
.map(function (x) {
  return x.type === 'Str' ? x.raw : extractText(x);
}
```

If not, what's special about links in particular?

File: lib/transform.js [hunk 3]
```diff
@@ -136,8 +114,7 @@ exports = module.exports = function transform(content, mode, maxHeaderLevel, tit
   var currentToc = info.hasStart && lines.slice(info.startIdx, info.endIdx).join('\n')
     , linesToToc = getLinesToToc(lines, currentToc, info);
 
-  var headers = getHashedHeaders(linesToToc, maxHeaderLevel)
-    .concat(getUnderlinedHeaders(linesToToc, maxHeaderLevel))
+  var headers = getMarkdownHeaders(linesToToc, maxHeaderLevel)
     .concat(getHtmlHeaders(linesToToc, maxHeaderLevelHtml))
```
Comment (jez) [inline lib/transform.js]: Do we still need to parse HTML headers ourselves? I would have hoped that `markdown-to-ast` would parse that for us, but maybe not.

If it doesn't, do we want to do anything special to handle the case of `<h1><a href="#">Header</a><h1>`? I'm fine with not supporting this edge case.

**Comment by jez** [comment]
I left a few comments in line, mostly just curious about how things work. I think this PR is really cool!

**Comment by thlorenz** [comment]
Good idea to move away from the original simplistic header parser.
We may still need to parse out headers of html manually afaik.

I'm wondering like @jez if this is the only markdown parser that gives us what we need. Specifically I know that marked supports handling sections like html separately. Mainly I'm just interested with your experiences with markdown-to-ast.

It'd be totally rad if we could fix the html header parser as well. Currently it'll pick up html inside code sections (vs. actual inlined html).
Any thoughts on that?

Overall like this PR as well especially since all tests still pass unmodified.
So good from my end once @jez's concerns were addressed.

File: package.json [hunk 4]
```diff
@@ -18,6 +18,7 @@
   "dependencies": {
     "anchor-markdown-header": "^0.5.4",
     "htmlparser2": "~3.7.1",
+    "markdown-to-ast": "^3.0.5",
```
Comment (thlorenz) [inline package.json]: Please stick to `~` ranges for now.

File: lib/transform.js (hunk 1 above)
Comment (jmanuel1) [inline lib/transform.js]: I saw line numbers are used to order headers, and `markdown-to-ast` gives location info. I search on Google and Github for "markdown parsers" and "markdown ast" and used `markdown-to-ast` because it was simpler to use. I did see other options; `marked-ast` doesn't give locations. As for `markdown-ast`... should I try it?

File: lib/transform.js (hunk 2 above)
Comment (jmanuel1) [inline lib/transform.js]: I'm thinking `code spans`, *emphasis*, and **strong** should be kept. I'm not sure how the parser acts with autolinks.

Should we handle the case there is html in a header?

File: lib/transform.js (hunk 3 above)
Comment (jmanuel1) [inline lib/transform.js]: I just wanted to get the PR out so you could take a look at it. `markdown-to-ast` will give ast nodes to HTML, but it won't parse it.

Comment (thlorenz) [inline lib/transform.js]: OK, so I suppose we still have to work on fixing issues like #36.

File: lib/transform.js (hunk 1 above)
Comment (jez) [inline lib/transform.js]: No, I was mostly just curious. `markdown-to-ast` looks fine. I only really asked because I know that `marked` boasts about how fast it is, but It looks like you have to do a bit of work to expose the underlying AST, and if it still doesn't give line numbers like you mention, then it's probably not worth it.

File: lib/transform.js (hunk 2 above)
Comment (jez) [inline lib/transform.js]: Oh hmm, I was thinking none of those should be kept. I guess I'll defer to @thlorenz then.



> Should we handle the case there is html in a header?

I don't think that many people have run into this issue. Mostly people have additional markdown within their headers. We can potentially address this if at some point we decide to solidify our story around "parsing" HTML in the first place.

**Comment by thlorenz** [comment]
LGTM overall. Let's squash this and merge to master.
We can address any other issues in another PR.
Unless you have an objection @jez

**Comment by jez** [comment]
:+1:

**Comment by thlorenz** [comment]
@jmanuel1 could you please squash these into one commit so we can merge to master?

**Comment by thlorenz** [comment]
@jez @jmanuel1 not sure why this became dormant.
I believe we got some related issues that would most likely be fixed when this gets pulled in, i.e. https://github.com/thlorenz/doctoc/issues/86

Could we get this squashed and merged? If neither of you have the time to do it I'll try to get to it this or next week.

**Comment by jmanuel1** [comment]
OK, I've figured out squashing. All squashed!

**Comment by thlorenz** [comment]
Tried to finally merge this but am running into a test failure:

```
➝  node -v
v4.1.2

➝  npm ts

> doctoc@0.15.0 test /Users/thlorenz/dev/js/projects/doctoc
> set -e; for t in test/*.js; do node $t; done

TAP version 13

# given a file that includes html with header tags and maxHeaderLevel 8
ok 1 generates correct toc for non html and html headers

# given a file that includes html with header tags using default maxHeaderLevel
ok 2 generates correct toc for non html and html headers omitting headers larger than maxHeaderLevel

1..2
# tests 2
# pass  2

# ok
TAP version 13

# overwrite existing title
ok 1 generates correct toc for when custom --title was passed

# do not overwrite existing title
ok 2 generates correct toc for when custom title exists in README already

# clobber existing title
ok 3 generates correct toc for when --notitle was passed

1..3
# tests 3
# pass  3

# ok
TAP version 13

# given a file with edge-case header names
ok 1 generates a correct toc when headers are weird

1..1
# tests 1
# pass  1

# ok
TAP version 13
# transforming
ok 1 transforms it
ok 2 generates correct anchors
# transforming
ok 3 transforms it
ok 4 generates correct anchors
# transforming
ok 5 transforms it
ok 6 generates correct anchors
# transforming
ok 7 transforms it
ok 8 generates correct anchors
# transforming
ok 9 transforms it
ok 10 generates correct anchors
# transforming
ok 11 transforms it
ok 12 generates correct anchors
# transforming
ok 13 transforms it
ok 14 generates correct anchors
# transforming
ok 15 transforms it
ok 16 generates correct anchors
# transforming
ok 17 transforms it
ok 18 generates correct anchors
# transforming
ok 19 transforms it
ok 20 generates correct anchors
# transforming
ok 21 transforms it
ok 22 generates correct anchors
# transforming
ok 23 transforms it
ok 24 generates correct anchors
# transforming
ok 25 transforms it
ok 26 generates correct anchors
# transforming when old toc exists
not ok 27 transforms it
  ---
    file:   /Users/thlorenz/dev/js/projects/doctoc/test/transform.js
    line:   242
    column: 5
    stack:
      - |
        getCaller (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-assert.js:418:17)
      - |
        Function.assert (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-assert.js:21:16)
      - |
        Test._testAssert [as ok] (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-test.js:87:16)
      - |
        Test.<anonymous> (/Users/thlorenz/dev/js/projects/doctoc/test/transform.js:242:5)
      - |
        emitNone (events.js:72:20)
      - |
        Test.emit (events.js:166:7)
      - |
        Test.emit (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-test.js:104:8)
      - |
        GlobalHarness.Harness.process (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-harness.js:87:13)
      - |
        doNTCallback0 (node.js:408:9)
      - |
        process._tickCallback (node.js:337:13)
  ...
not ok 28 TypeError: Cannot read property 'split' of undefined
  ---
    type:    TypeError
    message: Cannot read property 'split' of undefined
    code:    ~
    errno:   ~
    file:    /Users/thlorenz/dev/js/projects/doctoc/test/transform.js
    line:    245
    column:  14
    stack:
      - |
        Test.<anonymous> (/Users/thlorenz/dev/js/projects/doctoc/test/transform.js:245:14)
      - |
        emitNone (events.js:72:20)
      - |
        Test.emit (events.js:166:7)
      - |
        Test.emit (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-test.js:104:8)
      - |
        GlobalHarness.Harness.process (/Users/thlorenz/dev/js/projects/doctoc/node_modules/tap/lib/tap-harness.js:87:13)
      - |
        doNTCallback0 (node.js:408:9)
      - |
        process._tickCallback (node.js:337:13)
      - |
        Function.Module.runMain (module.js:477:11)
      - |
        startup (node.js:118:18)
      - |
        node.js:952:3
    thrown:  true
  ...
# transforming
ok 29 transforms it
ok 30 generates correct anchors
# transforming
ok 31 transforms it
ok 32 generates correct anchors

1..32
# tests 32
# pass  30
# fail  2
```

Do you know what's going on here @jmanuel1?
